### PR TITLE
Add some documentation of the perceived problems.

### PR DESCRIPTION
**Pull request by mjtamlyn**
Longer description of the motivation behind the project. I found it helpful to speak less about the specifics and more about the wider story which explains why we have what we have, and why it's a bit weak. I have a tendency to be very wordy, suggestions of where this section can be shortened without losing much detail are welcome.

TODO:
- [ ] Expand the "good parts" section

### CODE REVIEW

File: docs/source/problems.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Django needs to recognise that the web platform has moved on. It speaks
+multiple encodings and protocols, and we must make it easy for developers to
+build their websites independently of their current choice of encoding. In
+particular, the design of core features (such as validation of incoming data)
+should not be oriented strongly around one particular encoding. Django should
+also have a built in system for translating from the database to an encodable
+representation of that data.
+
+Your database structure does match the resources you expose
```
Comment (phalt) [inline docs/source/problems.rst]: `does not`*

**Comment by phalt** [comment]
I think this is good progress, and being wordy isn't a bad thing in this particular case, as there are many different reasons for this change, and already many different fragmented solutions.

File: docs/source/problems.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Introspectability
+=================
+
+If you wish to have a shared engine for data input and output but interact with
+it in different ways, then you need to ensure that the structure of that engine
+is easily introspectable. Whilst Django forms and DRF serializers do document
+some of their internal structure, it is not structured in a way which is easily
+serializable to be sent to a client application, or inspected by different
+rending platforms. By considering introspection as a top priority, we can make
```
Comment (Ian-Foote) [inline docs/source/problems.rst]: `rendering`

File: docs/source/problems.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+it easier for an ecosystem of libraries to consume this to build self
+describing APIs, browser rendered forms, and mirrored validation. It is
```
Comment (Ian-Foote) [inline docs/source/problems.rst]: `self-describing`?

File: docs/source/problems.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+important that not only the core field names and types themselves is
+instrospectable, but also the relationships between them for validation and
+display logic.
+
+The good parts
+==============
+
+The declative API
```
Comment (Ian-Foote) [inline docs/source/problems.rst]: `declarative`

File: docs/source/index.rst [hunk 5]
```diff
@@ -1,22 +1,41 @@
...
-   sphinx-quickstart on Sun Apr  3 09:38:05 2016.
-   You can adapt this file completely to your liking, but it should at least
-   contain the root `toctree` directive.
-
-Welcome to Django serial forms's documentation!
-===============================================
-
-Contents:
+===================
+Django serial forms
```
Comment (MoritzS) [inline docs/source/index.rst]: Django adapters ;-)

File: docs/source/problems.rst [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+communication architecture will become the new standard.
+
+Fundamentally, a data driven website such as the ones typically built using
+Django handles two major external processes. It delivers a representation of
+some data, and it receives requests to modify that data. In the current
+ecosystem, there are two distinct libraries most commonly used to achieve these
+core goals, and developers may mix and match between them within their
+projects. The first is core Django, which excels at rendering rich data into a
+document, and at accepting HTML form encoded data, validating it and storing
+it. The second is Django Rest Framework, which is optimised for building a
```
Comment (MoritzS) [inline docs/source/problems.rst]: optimized

File: docs/source/problems.rst [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+developer friendly API, generally encoded in JSON. These two codebases both do
+a good job of their work, but the choice between them for a developer is made
+based on how you want to construct the architecture of your website. If you
+have built an HTML oriented website and then wish to add an API, you have to
+redefine many core parts of your logic in a similar bit subtly different
```
Comment (MoritzS) [inline docs/source/problems.rst]: bit -> but

File: docs/source/problems.rst [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+system.
+
+Django needs to recognise that the web platform has moved on. It speaks
```
Comment (MoritzS) [inline docs/source/problems.rst]: recognize

File: docs/source/problems.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+also have a built in system for translating from the database to an encodable
+representation of that data.
+
+Your database structure does match the resources you expose
+===========================================================
+
+The design of Django forms (and model forms) came out of a desire to easily
+provide an interface to an underlying relational database. A relational
+database stores a set of fields together as a single record, and the table can
+be desribed as the collection of those fields, with a few extra conditions.
```
Comment (MoritzS) [inline docs/source/problems.rst]: desribed -> described

File: docs/source/problems.rst [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+Naturally, the design of forms copied this, and a form is a collection of
+fields.
+
+Over time, almost all Django sites will start to diverge in some way from
```
Comment (MoritzS) [inline docs/source/problems.rst]: start instead of will start?

File: docs/source/problems.rst [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+operate on related parent or child tables as if they are a single resource,
+they may represent only certain parts of the tables, and they may expose the
+same data in different ways depending on the user or application requesting it.
+
+Both Django forms and DRF serializers allow you to change them by inheritance,
+using a metaclass to organise the set of fields, and python inheritance for any
+cross-field needs. Django forms provide no easy way to nest resources, or easy
+ways to reuse the same logical fields with different rendering rules. DRF
+serializers provide more limited ways to include cross field logic. Both
+provide a model oriented version which reduces the amount of work and
```
Comment (MoritzS) [inline docs/source/problems.rst]: and -> an

File: docs/source/problems.rst [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+individual developer needs to do to interact with a single model, but the
+adaptability of these are limited when interacting with multiple models, often
+resulting in users needing to duplicate code, or reimplement the model binding.
+The inheritance pattern also makes it easy to add pieces, but harder to cut
+them up, making interfaces where you can read or write only certain parts of the
+data harder to build.
```
Comment (MoritzS) [inline docs/source/problems.rst]: This sentence sounds strange, especially the "harder to build" at the end.

File: docs/source/problems.rst (hunk 2 above)
Comment (MoritzS) [inline docs/source/problems.rst]: Where would you put "rendering" here? I don't see where it would fit.

File: docs/source/problems.rst [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+If you wish to have a shared engine for data input and output but interact with
+it in different ways, then you need to ensure that the structure of that engine
+is easily introspectable. Whilst Django forms and DRF serializers do document
+some of their internal structure, it is not structured in a way which is easily
+serializable to be sent to a client application, or inspected by different
+rending platforms. By considering introspection as a top priority, we can make
+it easier for an ecosystem of libraries to consume this to build self
+describing APIs, browser rendered forms, and mirrored validation. It is
+important that not only the core field names and types themselves is
```
Comment (MoritzS) [inline docs/source/problems.rst]: is -> are

File: docs/source/problems.rst (hunk 2 above)
Comment (Ian-Foote) [inline docs/source/problems.rst]: Instead of `rending`.

Comment (MoritzS) [inline docs/source/problems.rst]: You're right, I completely missed that!

File: docs/source/problems.rst (hunk 8 above)
Comment (mjtamlyn) [inline docs/source/problems.rst]: 🇬🇧 🇬🇧 🇬🇧

Comment (phalt) [inline docs/source/problems.rst]: @mjtamlyn on a serious note, what version of English are we choosing to write this in? Django docs are in American English?

Comment (mjtamlyn) [inline docs/source/problems.rst]: Yes, it will be in 🇺🇸 , I'm just making my excuses for being 🇬🇧 !

**Pull request closed without merging**